server/src/main/java/io/druid/server/namespace/NamespacedExtractionModule.java

```diff
@@ -0,0 +1,448 @@
+              final String sData = StringUtils.fromUtf8(data);
+              final ExtractionNamespace namespace = jsonMapper.readValue(sData, ExtractionNamespace.class);
+              final String ns = namespace.getNamespace();
+              final PathChildrenCacheEvent.Type eventType = pathChildrenCacheEvent.getType();
+              log.debug("Processing event type [%s]", eventType);
+
+              if (eventType.equals(PathChildrenCacheEvent.Type.CHILD_REMOVED) ||
+                  eventType.equals(PathChildrenCacheEvent.Type.CHILD_UPDATED)) {
+                log.debug("Removing namespace [%s]", ns);
+                namespaceExtractionCacheManager.delete(ns);
+              }
+
+              if (eventType.equals(PathChildrenCacheEvent.Type.CHILD_ADDED) ||
+                  eventType.equals(PathChildrenCacheEvent.Type.CHILD_UPDATED)) {
+                log.debug("Adding namespace [%s] : %s", ns, namespace.toString());
+                final ExtractionNamespaceFunctionFactory factory = injector.getInstance(
+                    Key.get(
+                        ExtractionNamespaceFunctionFactory.class,
+                        Names.named(
+                            namespace
+                                .getClass()
+                                .getCanonicalName()
+                        )
+                    )
+                );
+
+                log.info(
+                    "New Update rule for namespace [%s]",
+                    namespace.toString()
+                );
+                namespaceExtractionCacheManager.schedule(namespace, factory);
+              }
+            }
+          };
+      pathChildrenCache.getListenable().addListener(
+          listener,
+          listeningExecutorService
+      );
+      try {
+        pathChildrenCache.start();
+      }
+      catch (Exception e) {
+        throw Throwables.propagate(e);
+      }
+    }
+
+    @LifecycleStop
+    public void stop()
+    {
+      Exception exception = null;
+      try {
+        pathChildrenCache.getListenable().removeListener(listener);
+      } catch (Exception e){
+        exception = e;
+      }
+      listener = null;
+
+      try {
+        pathChildrenCache.close();
+      }
+      catch (IOException e) {
+        if(exception == null){
+          exception = e;
+        } else {
+          exception.addSuppressed(e);
+        }
+      }
+      try {
+        curator.close();
+      }catch(Exception e){
+        if(exception == null){
+          exception = e;
+        } else {
+          exception.addSuppressed(e);
+        }
+      }
+      if(exception != null){
+        throw Throwables.propagate(exception);
+      }
+    }
+
+    public Collection<String> listNamespaces()
+    {
+      final ArrayList<Exception> innerExceptions = new ArrayList<>();
+      Set<String> retval = null;
+      try {
+        retval = ImmutableSet.copyOf(
+            Collections2.filter(
```

Comment:
Between here and the resource, you are reaching in and grabbing things from the base persistent storage in multiple places.  It would be better to create an object that does the talking to the persistent store and use that instead.